### PR TITLE
Adds VLM Training support to SFTTrainer + VSFT script

### PR DESCRIPTION
**Pull request by edbeeching**
- Modifies the SFTTrainer so that it can be used with a VLM SFT dataset
- Adds an example script `vsft.py` to train the llava1.5 model with an instruct dataset

TODO:
- [x] Test PEFT support
- [x] Run full training
- [x] Add example to docs
- [x] add tests

Example usage:
```
python examples/scripts/vsft.py \
    --model_name_or_path="llava-hf/llava-1.5-7b-hf" \
    --report_to="wandb" \
    --learning_rate=1.4e-5 \
    --per_device_train_batch_size=8 \
    --gradient_accumulation_steps=1 \
    --output_dir="data/vsft-llava-1.5-7b-hf" \
    --logging_steps=5 \
    --num_train_epochs=1 \
    --push_to_hub \
    --gradient_checkpointing \
    --remove_unused_columns=False \
    --torch_dtype=float16 \
    --fp16=True \ 
    --dataset_name=HuggingFaceH4/llava-instruct-mix-vsft \
```

### CODE REVIEW

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:20]: Maybe pick something a bit less magic?
```suggestion
    --learning_rate=1.00e-5 \
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:93]: For gradient checkpointing I think we need to manually add the `use_reentrant=False` flag like we do here: https://github.com/huggingface/trl/blob/57aebe9c36021e679662181468b04ff42432daf3/examples/scripts/reward_modeling.py#L46C5-L46C76

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:104]: Should this be configurable from the command line instead of hardcoded? I think it's also not quite correct since we're missing the system prompt which is quite important for evals. The full template is something like the following, adjusted for your new data schema:

```
"{% for message in messages %}{% if loop.index0 == 0 %}A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. USER: {{ message['content'] }} {% elif message['role'] == 'user' %}USER: {{ message['content'] }} {% else %} ASSISTANT: {{ message['content'] }}{{ eos_token }}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ 'ASSISTANT:' }}{% endif %}"
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:117]: Do you mean llava doesn't have a cache support or that gradient checkpointing isn't supported? If the latter, then we should remove it from the command line args

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:122]: Delete?

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:133]: Should this live in a `collators.py` module? That way we can properly unit test it (which would be good in any case)

File: trl/trainer/sft_trainer.py
Comment (lewtun) [inline trl/trainer/sft_trainer.py:405]: Let's add an integration test for this case under the existing `SFTTrainer` tests, along with an example training a tiny random llava model

**Review by lewtun — COMMENTED** [review]
Very clean implementation @edbeeching ! Just a few nits and a few unit/integration tests needed and this should be good to merge

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:38]:
```suggestion
    --learning_rate=1.00e-4 \
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:44]:
```suggestion
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:26]:
```suggestion
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:24]: Are you sure we want to log every step? Maybe 
```suggestion
    --logging_steps=5 \
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:42]:
```suggestion
    --logging_steps=5 \
```

File: examples/scripts/vsft.py
Comment (lewtun) [inline examples/scripts/vsft.py:198]: After this I think we might need a final `trainer.push_to_hub()`

File: examples/scripts/vsft.py
Comment (pcuenca) [reply examples/scripts/vsft.py:104]: Could this (or parts of it) be potentially derived from the tokenizer, if a chat template is defined for the model?

Comment (pcuenca) [reply examples/scripts/vsft.py:104]: Also, if this is Llava-specific, I'd call the script `vsft_llava.py` (or rename this variable to `CHAT_TEMPLATE`).

File: examples/scripts/vsft.py
Comment (pcuenca) [reply examples/scripts/vsft.py:133]: For an initial example I'd keep it simple and self-contained. I think it's valuable to understand everything and easy to tweak when necessary.

File: examples/scripts/vsft.py
Comment (lewtun) [reply examples/scripts/vsft.py:104]: > Could this (or parts of it) be potentially derived from the tokenizer, if a chat template is defined for the model?

Indeed, ideally we should use the tokenizer/processor chat template if it exists and then fallback to the default llava chat template if needed.

File: examples/scripts/vsft.py
Comment (edbeeching) [reply examples/scripts/vsft.py:20]: Will do, the value is copied from the SFT example: https://github.com/huggingface/trl/blob/57aebe9c36021e679662181468b04ff42432daf3/examples/scripts/sft.py#L20

File: examples/scripts/vsft.py
Comment (edbeeching) [reply examples/scripts/vsft.py:117]: It has this option is the base LLM, but the LlavaForConditionalGeneration model does not expose it.

File: examples/scripts/vsft.py
Comment (edbeeching) [reply examples/scripts/vsft.py:104]: Thanks, I added the system prompt. 

There are a lot of llava specific things in this script, the model itself, the chat template, etc. I am wondering if it makes sense to make it more general now, or to wait until we can support more VLM models?

Comment (pcuenca) [reply examples/scripts/vsft.py:104]: In my opinion, I'd keep it llava specific for the release so it's as easy to understand as possible. We could invite the community to open PRs for other examples and then decide whether to write a more general version.

File: examples/scripts/vsft_llava.py
Comment (lewtun) [inline examples/scripts/vsft_llava.py:52]:
```suggestion

# to evaluate, first install the lmms-eval framework: pip install git+https://github.com/EvolvingLMMs-Lab/lmms-eval.git
# then run:
accelerate launch --num_processes=8 -m lmms_eval \
        --model llava_hf \
        --model_args pretrained=llava-hf/llava-1.5-7b-hf \
        --tasks mmbench \
        --batch_size 1 \
        --output_path ./logs/ \
        --log_sample
"""
```

**Review by younesbelkada — APPROVED** [review]
Thanks ! We should be good to go after fixing the merge conflict on main ! 🚀

**Pull request merged**
